lib/puppetserver/ca/setup_command.rb

```diff
@@ -0,0 +1,136 @@
+require 'optparse'
+require 'puppetserver/ca/x509_loader'
+require 'puppetserver/ca/puppet_config'
+
+module Puppetserver
+  module Ca
+    class SetupCommand
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run!(input)
+        bundle_path = input['cert-bundle']
+        key_path = input['private-key']
+        chain_path = input['crl-chain']
+        config_path = input['config']
+
+        files = [bundle_path, key_path, chain_path, config_path].compact
+
+        errors = validate_file_paths(files)
+        return 1 if log_possible_errors(errors)
+
+        unless chain_path
+          @logger.err 'Warning:'
+          @logger.err '    No CRL chain given'
+          @logger.err '    Full CRL chain checking will not be possible'
+          @logger.err ''
+        end
+
+        loader = X509Loader.new(bundle_path, key_path, chain_path)
+        return 1 if log_possible_errors(loader.errors)
+
+        puppet = PuppetConfig.parse(config_path)
+        return 1 if log_possible_errors(puppet.errors)
+
+
+        File.open(puppet.settings[:cacert], 'w') do |f|
+          loader.certs.each do |cert|
+            f.puts cert.to_pem
+          end
+        end
+
+        File.open(puppet.settings[:cakey], 'w') do |f|
+          f.puts loader.key.to_pem
+        end
+
+        File.open(puppet.settings[:cacrl], 'w') do |f|
+          loader.crls.each do |crl|
+            f.puts crl.to_pem
+          end
+        end
+
+        return 0
+      end
+
+      def log_possible_errors(maybe_errors)
+        errors = Array(maybe_errors).compact
+        unless errors.empty?
+          @logger.err "Error:"
+          errors.each do |message|
+            @logger.err "    #{message}"
+          end
+          return true
+        end
+      end
+
+      def parse(cli_args)
+        parser, inputs = parse_inputs(cli_args)
+        exit_code = validate_inputs(inputs, parser.help)
+
+        return inputs, exit_code
+      end
+
+      def validate_inputs(input, usage)
+        exit_code = nil
+
+        if input['cert-bundle'].nil? || input['private-key'].nil?
+          @logger.err 'Error:'
+          @logger.err 'Missing required argument'
+          @logger.err '    Both --cert-bundle and --private-key are required'
+          @logger.err ''
+          @logger.err usage
```

Comment:
I notice we use @logger a lot.  Would there be any advance/disadvantage to using an ```attr_reader :logger?```

Reply:
I used the instance variable directly since there's no external reader of the logger, and we don't mock out the logger in the tests (and if we did that in a test of SetupCommand, we could pass the logger into the instance under test).

Having said, that `attr_reader :logger` might be more idiomatic?